File: backend/pkg/models/overtime_request.go
package models

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
	"time"
)

// ==========================================
// Subdocument for reward info
type OvertimeReward struct {
	RewardType    string     `json:"reward_type" bson:"reward_type"`                     // money|time_off
	RewardOption  string     `json:"reward_option,omitempty" bson:"reward_option,omitempty"` // early_out|late_in
	RewardDate    *time.Time `json:"reward_date,omitempty" bson:"reward_date,omitempty"`   // Tanggal klaim reward (terutama untuk time_off)
	RewardNominal float64    `json:"reward_nominal,omitempty" bson:"reward_nominal,omitempty"` // Nominal uang lembur
	Status        string     `json:"status" bson:"status"`                               // none|pending|granted|used
	GrantedAt     *time.Time `json:"granted_at,omitempty" bson:"granted_at,omitempty"`
	UsedAt        *time.Time `json:"used_at,omitempty" bson:"used_at,omitempty"`
}

// ==========================================
// Subdocument for each employee
type OvertimeEmployee struct {
	UserID         primitive.ObjectID `json:"user_id" bson:"user_id"`
	EmployeeStatus string             `json:"employee_status" bson:"employee_status"`                   // pending|agreed|rejected
	RejectionNote  string             `json:"rejection_note,omitempty" bson:"rejection_note,omitempty"` // opsional, wajib jika rejected
	LetterURL      string             `json:"letter_url,omitempty" bson:"letter_url,omitempty"`         // link SPKL per karyawan
	ConfirmedAt    *time.Time         `json:"confirmed_at,omitempty" bson:"confirmed_at,omitempty"`
	Reward         OvertimeReward     `json:"reward" bson:"reward"`
}

// ==========================================
type OvertimeRequest struct {
	ID            primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	DepartmentID  primitive.ObjectID `json:"department_id" bson:"department_id"`     // pengaju
	RequestedByID primitive.ObjectID `json:"requested_by_id" bson:"requested_by_id"` // kepala departemen yang submit
	Date          time.Time          `json:"date" bson:"date"`                       // hanya tanggal, jam di bawah
	StartTime     string             `json:"start_time" bson:"start_time"`           // format "HH:mm"
	EndTime       string             `json:"end_time" bson:"end_time"`               // format "HH:mm"
	Reason        string             `json:"reason" bson:"reason"`
	Status        string             `json:"status" bson:"status"`                             // draft|submitted|published
	Notes         string             `json:"notes,omitempty" bson:"notes,omitempty"`           // catatan HR opsional
	LetterURL     string             `json:"letter_url,omitempty" bson:"letter_url,omitempty"` // file/link SPKL opsional
	CreatedAt     time.Time          `json:"created_at" bson:"created_at"`
	UpdatedAt     time.Time          `json:"updated_at" bson:"updated_at"`

	Employees []OvertimeEmployee `json:"employees" bson:"employees"` // array karyawan lembur
}

func (r *OvertimeRequest) GetDurationHours() float64 {
	start, err1 := time.Parse("15:04", r.StartTime)
	end, err2 := time.Parse("15:04", r.EndTime)
	if err1 != nil || err2 != nil {
		return 0
	}
	duration := end.Sub(start)
	if duration < 0 {
		duration += 24 * time.Hour
	}
	return duration.Hours()
}

// ─── Constants ────────────────────────────────────────────────────────────

const (
	StatusDraft     = "draft"
	StatusSubmitted = "submitted"
	StatusPublished = "published"

	EmployeeStatusPending  = "pending"
	EmployeeStatusAgreed   = "agreed"
	EmployeeStatusRejected = "rejected"

	OvertimeRewardTypeMoney   = "money"
	OvertimeRewardTypeTimeOff = "time_off"

	OvertimeRewardStatusNone    = "none"
	OvertimeRewardStatusPending = "pending"
	OvertimeRewardStatusGranted = "granted"
	OvertimeRewardStatusUsed    = "used"
)

// ─── Request Models ───────────────────────────────────────────────────────

type OvertimeEmployeeInput struct {
	UserID string `json:"user_id" binding:"required"`
}

type CreateOvertimeRequestRequest struct {
	DepartmentID string                  `json:"department_id" binding:"required"`
	Date         string                  `json:"date" binding:"required"` // YYYY-MM-DD
	StartTime    string                  `json:"start_time" binding:"required"`
	EndTime      string                  `json:"end_time" binding:"required"`
	Reason       string                  `json:"reason" binding:"required"`
	Status       string                  `json:"status,omitempty"`
	Employees    []OvertimeEmployeeInput `json:"employees" binding:"required"`
}

type UpdateOvertimeRequestRequest struct {
	Date      *string                  `json:"date,omitempty"`
	StartTime *string                  `json:"start_time,omitempty"`
	EndTime   *string                  `json:"end_time,omitempty"`
	Reason    *string                  `json:"reason,omitempty"`
	Status    *string                  `json:"status,omitempty"`
	Employees *[]OvertimeEmployeeInput `json:"employees,omitempty"`
	Notes     *string                  `json:"notes,omitempty"`
	LetterURL *string                  `json:"letter_url,omitempty"`
}

type UpdateEmployeeStatusRequest struct {
	Status        string `json:"status" binding:"required"` // agreed|rejected
	RejectionNote string `json:"rejection_note,omitempty"`
}

type ClaimOvertimeRewardRequest struct {
	RewardType   string `json:"reward_type" binding:"required"` // money|time_off
	RewardOption string `json:"reward_option,omitempty"`        // early_out|late_in
	RewardDate   string `json:"reward_date,omitempty"`          // YYYY-MM-DD
}

// ─── Response Models ──────────────────────────────────────────────────────

type OvertimeEmployeeResponse struct {
	UserID         string         `json:"user_id"`
	FullName       string         `json:"full_name"`
	PayrollNumber  string         `json:"payroll_number"`
	PositionName   string         `json:"position_name"`
	EmployeeStatus string         `json:"employee_status"`
	RejectionNote  string         `json:"rejection_note,omitempty"`
	LetterURL      string         `json:"letter_url,omitempty"`
	ConfirmedAt    *time.Time     `json:"confirmed_at,omitempty"`
	Reward         OvertimeReward `json:"reward"`
}

type OvertimeRequestResponse struct {
	ID              string                     `json:"id"`
	DepartmentID    string                     `json:"department_id"`
	DepartmentName  string                     `json:"department_name"`
	RequestedByID   string                     `json:"requested_by_id"`
	RequestedByName string                     `json:"requested_by_name"`
	Date            time.Time                  `json:"date"`
	StartTime       string                     `json:"start_time"`
	EndTime         string                     `json:"end_time"`
	Reason          string                     `json:"reason"`
	Status          string                     `json:"status"`
	Notes           string                     `json:"notes,omitempty"`
	LetterURL       string                     `json:"letter_url,omitempty"`
	CreatedAt       time.Time                  `json:"created_at"`
	UpdatedAt       time.Time                  `json:"updated_at"`
	Employees       []OvertimeEmployeeResponse `json:"employees"`
}

// Legacy support or internal helper
type OvertimeApprovalResponse struct {
	Overtime OvertimeRequestResponse `json:"overtime"`
}

File: mobile_app/lib/models/overtime_request.dart
// lib/models/overtime_request.dart

// NEW FLOW: Kepala departemen mengajukan lembur untuk karyawan mereka
// Employees kemudian bisa agree/reject
// HR bisa publish surat

class OvertimeRequest {
  final String id;
  final String departmentId;
  final String departmentName;
  final String requestedById; // Kepala departemen yang submit
  final String requestedByName; // Display name
  final DateTime date;
  final String startTime; // format HH:mm
  final String endTime; // format HH:mm
  final String reason;
  final String status; // draft|submitted|published
  final String? notes;
  final String? letterUrl;
  final List<OvertimeEmployee> employees;
  final DateTime createdAt;
  final DateTime updatedAt;

  OvertimeRequest({
    required this.id,
    required this.departmentId,
    required this.departmentName,
    required this.requestedById,
    required this.requestedByName,
    required this.date,
    required this.startTime,
    required this.endTime,
    required this.reason,
    required this.status,
    this.notes,
    this.letterUrl,
    required this.employees,
    required this.createdAt,
    required this.updatedAt,
  });

  factory OvertimeRequest.fromJson(Map<String, dynamic> json) {
    return OvertimeRequest(
      id: json['id']?.toString() ?? '',
      departmentId: json['department_id']?.toString() ?? '',
      departmentName: (json['department_name'] ?? json['departmentName'] ?? '')
          .toString(),
      requestedById: json['requested_by_id']?.toString() ?? '',
      requestedByName:
          (json['requested_by_name'] ??
                  json['requestedByName'] ??
                  json['requested_by_id'] ??
                  '')
              .toString(),
      date: _parseDate(json['date']),
      startTime: json['start_time']?.toString() ?? '',
      endTime: json['end_time']?.toString() ?? '',
      reason: json['reason']?.toString() ?? '',
      status: (json['status'] ?? 'draft').toString().toLowerCase(),
      notes: json['notes']?.toString(),
      letterUrl: json['letter_url']?.toString(),
      employees:
          (json['employees'] as List?)
              ?.whereType<Map<String, dynamic>>()
              .map(OvertimeEmployee.fromJson)
              .toList() ??
          [],
      createdAt: _parseDate(json['created_at']),
      updatedAt: _parseDate(json['updated_at']),
    );
  }

  static DateTime _parseDate(dynamic v) {
    if (v == null) return DateTime.now();
    try {
      return DateTime.parse(v.toString()).toLocal();
    } catch (_) {
      return DateTime.now();
    }
  }

  // Helper untuk UI
  String get statusDisplay {
    switch (status) {
      case 'draft':
        return 'Draft';
      case 'submitted':
        return 'Dikirim';
      case 'published':
        return 'Dipublikasikan';
      default:
        return 'Unknown';
    }
  }

  bool get isDraft => status == 'draft';
  bool get isSubmitted => status == 'submitted';
  bool get isPublished => status == 'published';

  // Get hour & minute from time string
  (int, int) getStartTimeHourMin() {
    try {
      final parts = startTime.split(':');
      return (int.parse(parts[0]), int.parse(parts[1]));
    } catch (_) {
      return (0, 0);
    }
  }

  (int, int) getEndTimeHourMin() {
    try {
      final parts = endTime.split(':');
      return (int.parse(parts[0]), int.parse(parts[1]));
    } catch (_) {
      return (0, 0);
    }
  }

  // Calculate duration in hours
  double getDurationHours() {
    try {
      final start = startTime.split(':');
      final end = endTime.split(':');
      final startMins = int.parse(start[0]) * 60 + int.parse(start[1]);
      final endMins = int.parse(end[0]) * 60 + int.parse(end[1]);
      return (endMins - startMins) / 60;
    } catch (_) {
      return 0;
    }
  }
}

class OvertimeEmployee {
  final String userId;
  final String userName; // Display name
  final String employeeStatus; // pending|agreed|rejected
  final String? rejectionNote;
  final String? letterUrl; // URL dokumen SPKL untuk karyawan ini
  final DateTime? confirmedAt;
  final OvertimeReward? reward;

  OvertimeEmployee({
    required this.userId,
    required this.userName,
    required this.employeeStatus,
    this.rejectionNote,
    this.letterUrl,
    this.confirmedAt,
    this.reward,
  });

  factory OvertimeEmployee.fromJson(Map<String, dynamic> json) {
    return OvertimeEmployee(
      userId: json['user_id']?.toString() ?? '',
      userName: (json['full_name'] ?? json['user_name'] ?? json['name'] ?? '')
          .toString(),
      employeeStatus: (json['employee_status'] ?? 'pending')
          .toString()
          .toLowerCase(),
      rejectionNote: json['rejection_note']?.toString(),
      letterUrl: json['letter_url']?.toString(),
      confirmedAt: json['confirmed_at'] != null
          ? DateTime.tryParse(json['confirmed_at'].toString())
          : null,
      reward: json['reward'] != null
          ? OvertimeReward.fromJson(json['reward'] as Map<String, dynamic>)
          : null,
    );
  }

  String get displayName => userName.isNotEmpty ? userName : userId;

  String get statusDisplay {
    switch (employeeStatus) {
      case 'pending':
        return 'Menunggu';
      case 'agreed':
        return 'Setuju';
      case 'rejected':
        return 'Tolak';
      default:
        return 'Unknown';
    }
  }

  bool get isPending => employeeStatus == 'pending';
  bool get isAgreed => employeeStatus == 'agreed';
  bool get isRejected => employeeStatus == 'rejected';
}

class OvertimeReward {
  final String rewardType; // money|time_off
  final String? rewardOption; // early_out|late_in
  final String status; // none|pending|granted|used
  final DateTime? rewardDate; // Tanggal klaim reward (terutama untuk time_off)
  final DateTime? grantedAt;
  final DateTime? usedAt;

  OvertimeReward({
    required this.rewardType,
    this.rewardOption,
    required this.status,
    this.rewardDate,
    this.grantedAt,
    this.usedAt,
  });

  factory OvertimeReward.fromJson(Map<String, dynamic> json) {
    return OvertimeReward(
      rewardType: json['reward_type']?.toString() ?? '',
      rewardOption: json['reward_option']?.toString(),
      status: json['status']?.toString() ?? 'none',
      rewardDate: json['reward_date'] != null
          ? DateTime.tryParse(json['reward_date'].toString())
          : null,
      grantedAt: json['granted_at'] != null
          ? DateTime.tryParse(json['granted_at'].toString())
          : null,
      usedAt: json['used_at'] != null
          ? DateTime.tryParse(json['used_at'].toString())
          : null,
    );
  }

  String get rewardTypeDisplay {
    switch (rewardType) {
      case 'money':
        return 'Uang Lembur';
      case 'time_off':
        return 'Jam Kerja Dipercepat';
      default:
        return 'Belum Dipilih';
    }
  }

  String get statusDisplay {
    switch (status) {
      case 'none':
        return 'Belum Dipilih';
      case 'pending':
        return 'Menunggu Persetujuan';
      case 'granted':
        return 'Diterima';
      case 'used':
        return 'Sudah Digunakan';
      default:
        return 'Unknown';
    }
  }
}
